juju/juju
# Fix provider context

The call context passed provider api calls was not using the recently added std context, and was as yet only updating mongo when a credential becomes invalid. This PR simplifies the context implementation using a std context value to hold the invalidation callback. It also wires up the invalidation callback to update dqlite and mongo.

A few keys points:
1. We were jumping through hoops to create a provider context for bootstrap, add-model, destroy, kill
We don't want/need use the invalidation callback for these since the model is either not there yet or is being destroyed anyway. And bootstrap, add-model etc still report the same error.
2. A helper method is used to create the validation callback off a facade context during the register operation. This allows mongo to be transparently switched out later.
3. The statepolicy makes some provider api calls, mainly to do constraints checks, AZ querying etc. It's unwieldy to get things set up the new way, and statepolicy is being totally rewritten anyway. It just means that adding an app or unit or machine when the credential is invalid reports the error but does not mark the model as suspended etc. But this will eventually happen when the a worker causes the next provider api call. Another PR could look to add a temporary fix.

Commit 1: replace the somewhat convoluted provider context implementation in `environs/context` with one which uses a standard context with an invalidate callback attached.
Commit 2: the `apiserver/common/credentialcommon` package had credential validation logic which can now be moved to the `domain/credential/service` package and use the new context.
Commit 3: update all the providers to use the new context
Commit 4: update apiserver facades to use the new context


## QA steps

bootstrap with bad credential
```
$ juju bootstrap aws foo
ERROR 
The provided credentials could not be validated and 
may not be authorized to carry out the request.
Ensure that your account is authorized to use the Amazon EC2 service and 
that you are using the correct access keys. 
These keys are obtained via the "Security Credentials"
page in the AWS console: operation error EC2: DescribeAccountAttributes, https response error StatusCode: 401, RequestID: ec813b0d-3322-4e2b-b11e-131eae56428c, api error AuthFailure: AWS was not able to validate the provided access credentials
```

add 2 machines to a model
force update a bad credential
```
$ juju update-credential aws juju --controller test --force
Credential invalid for:
  controller:
    listing instances: 
The provided credentials could not be validated and 
may not be authorized to carry out the request.
Ensure that your account is authorized to use the Amazon EC2 service and 
that you are using the correct access keys. 
These keys are obtained via the "Security Credentials"
page in the AWS console: operation error EC2: DescribeInstances, https response error StatusCode: 401, RequestID: c63db4c7-f34e-4ddc-acb4-8ee925c31b09, api error AuthFailure: AWS was not able to validate the provided access credentials
Failed models may require a different credential.
Use ‘juju set-credential’ to change credential for these models.
```

try to remove a machine
```
$ juju remove-machine 0
WARNING This command will perform the following actions:
will remove machine 0

Continue [y/N]? y
$ juju status
Model  Controller  Cloud/Region        Version      SLA          Timestamp       Notes
foo    test        aws/ap-southeast-2  4.0-beta1.1  unsupported  12:57:54+10:00  suspended since cloud credential is not valid

App     Version  Status  Scale  Charm   Channel  Rev  Exposed  Message
ubuntu  22.04    active      1  ubuntu  stable    24  no       

Unit       Workload  Agent  Machine  Public address  Ports  Message
ubuntu/0*  active    idle   1        13.211.130.201         

Machine  State    Address         Inst id              Base          AZ               Message
0        stopped  3.26.221.139    i-0df10a1b90dd61936  ubuntu@22.04  ap-southeast-2c  running
1        started  13.211.130.201  i-0e632f61246aed26f  ubuntu@22.04  ap-southeast-2a  running

$ juju show-model
foo:
  name: admin/foo
  short-name: foo
  model-uuid: 7c5c93f4-f678-4f71-8a56-c954ecf82054
  model-type: iaas
  controller-uuid: f1fa15ed-187d-4409-80ea-558bfaefd7a6
  controller-name: test
  is-controller: false
  owner: admin
  cloud: aws
  region: ap-southeast-2
  type: ec2
  life: alive
  status:
    current: suspended
    message: suspended since cloud credential is not valid
    reason: "The provided credentials could not be validated and \nmay not be authorized
      to carry out the request.\nEnsure that your account is authorized to use the
      Amazon EC2 service and \nthat you are using the correct access keys. \nThese
      keys are obtained via the \"Security Credentials\"\npage in the AWS console:
      operation error EC2: DescribeInstances, https response error StatusCode: 401,
      RequestID: a1fa5450-e62e-470a-ae6b-1881c0e3ad12, api error AuthFailure: AWS
      was not able to validate the provided access credentials"
...
  credential:
    name: juju
    owner: admin
    cloud: aws
    validity-check: invalid
```

update the credential to be valid again

```
$ juju update-credential aws juju --controller test
Credential valid for:
  controller
Controller credential "juju" for user "admin" for cloud "aws" on controller "test" updated.
For more information, see ‘juju show-credential aws juju’.

$ juju status
Model  Controller  Cloud/Region        Version      SLA          Timestamp
foo    test        aws/ap-southeast-2  4.0-beta1.1  unsupported  13:03:52+10:00

App     Version  Status  Scale  Charm   Channel  Rev  Exposed  Message
ubuntu  22.04    active      1  ubuntu  stable    24  no       

Unit       Workload  Agent  Machine  Public address  Ports  Message
ubuntu/0*  active    idle   1        13.211.130.201         

Machine  State    Address         Inst id              Base          AZ               Message
1        started  13.211.130.201  i-0e632f61246aed26f  ubuntu@22.04  ap-southeast-2a  running
$ juju show-model
foo:
  name: admin/foo
  short-name: foo
  model-uuid: 7c5c93f4-f678-4f71-8a56-c954ecf82054
  model-type: iaas
  controller-uuid: f1fa15ed-187d-4409-80ea-558bfaefd7a6
  controller-name: test
  is-controller: false
  owner: admin
  cloud: aws
  region: ap-southeast-2
  type: ec2
  life: alive
  status:
    current: available
    since: 45 minutes ago
...
  credential:
    name: juju
    owner: admin
    cloud: aws
    validity-check: valid
```

## Links

**Jira card:** JUJU-[4622]
